# Домашнє завдання до теми "Python: Цикли"
В цьому ДЗ знову важливою умовою виконання завдань є те, що ми не імпортуємо жодних бібліотек, а робимо все на чистому Python, використовуючи лише базовий функціонал. Це необхідно аби відпрацювати роботу з базовими типами даних і циклами. А вже з наступного модуля ми во всю будемо використовувати бібліотеки :)

0.Розігрів. Напишіть цикл, який виводить на екран числа від 1 до 10 включно.

In [1]:
for i in range(1, 11):
    print(i, end="; ")

1; 2; 3; 4; 5; 6; 7; 8; 9; 10; 

**Початок покладено!**

Для наступних завдань у нас є наступний список:

In [3]:
numbers = [42, 51, 29, 2, 13, 44, 80, 66, 16]

1.Використовуючи цикл for cтворіть новий список `numbers_squared`, де кожен елемент списку `numbers` є домноженим на 2. Нагадаю, що фактично коли ми з допомогою циклу отримуємо новий список, то ми спочатку створюємо пустий список, а потім додаємо в нього новий елемент на кожній ітерації. Новий елемент буде в даному випадку результатом домножання чергового елемента вихідного списку `numbers` на 2.

In [4]:
numbers_squared = []
for i in numbers:
    numbers_squared.append(i*2)
print(numbers_squared)

[84, 102, 58, 4, 26, 88, 160, 132, 32]


Очікуваний результат: `[84, 102, 58, 4, 26, 88, 160, 132, 32]`

2.Використовуючи `list comprehension` cтворіть новий список `numbers_squared_lh`, де кожен елемент списку `numbers` є домноженим 2.
Так, операція, яку ми викнуємо над кожним елементом вихідного списку, та сама, що і в попередньому завданні, але інший спосіб отримання результату.

In [9]:
numbers_squared_lh = [i*2 for i in numbers]
print(numbers_squared_lh)


[84, 102, 58, 4, 26, 88, 160, 132, 32]


Очікуваний результат: `[84, 102, 58, 4, 26, 88, 160, 132, 32]`

3.Задано два списки в клітинці нижче - `a` і `b`. Створіть цикл, результатом якого буде список `result`, що містить елементи типу bool і який фактично каже, 
"чи на цьому індексі елементи наших двох списків співпадають". Приклад:
```
a = [1, 2, 3, 6]
b = [1, 7, 3, 6]
# ми виконали поелементне порівняння двох списків і отримали новий список result
# result містить наступний результат -> [True, False, True, True]
```
В кінці виведіть вміст result на екран.

In [10]:
a = [64, 93, 60, 95, 66, 9, 15, 55, 77, 28, 74, 100, 13, 1, 85, 48, 66, 65, 59, 32]
b = [64, 93, 60, 95, 66, 9, 15, 56, 77, 28, 74, 100, 13, 1, 85, 48, 67, 65, 59, 32]

In [16]:
result = []
for i in range(len(a)):
    result.append(a[i]==b[i])
print(result)

[True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, False, True, True, True]


4.Давайте знайдемо середнє арифметичне значення елементів списку `numbers` округлене до 2 цифр після коми та запишемо його у змінну `numbers_mean`. Для цього виконаємо два простих кроки.
1. Знайдіть суму всіх елементів списку `numbers` та запишіть її у змінну `numbers_sum`.
2. Обчисліть `numbers_mean` шляхом ділення суми елементів у списку на їх кількість. Кількість елементів в списку - це його довжина.

Виведіть `numbers_mean`, округлений до 2 цифр після коми, на екран.

**Підказка**: округлити число можна з використанням вбудованої функції Python [round()](https://docs.python.org/3/library/functions.html#round).

In [6]:
numbers_sum = sum(numbers)
numbers_mean = numbers_sum/len(numbers)
round(numbers_mean, 2)

38.11

Очікуваний результат: `38.11`

5.Використовуючи цикл, створіть новий список `numbers_gt_mean`, який порівнює кожен елемент з середнім значенням за списокм, знайденим в попередньому завданні, за наступним правилом: "якщо цей елемент спиcку `numbers` є більшим за `numbers_mean`, то запишемо 1, інакше 0". Виведіть новий список на екран.

In [7]:
numbers_gt_mean = []
for i in range(len(numbers)):
    if numbers[i] > numbers_mean:
        numbers_gt_mean.append(1)
    else:
        numbers_gt_mean.append(0)
print(numbers_gt_mean)

[1, 1, 0, 0, 0, 1, 1, 1, 0]


Очікуваний результат: `[1, 1, 0, 0, 0, 1, 1, 1, 0]`

6.Отримайте список за тим самим правилом, що в завданні 5, тільки тепер з list comprehension. Результат запишіть у змінну `numbers_gt_mean_lc`

In [11]:
numbers_gt_mean_ls = [1 if numbers[i] > numbers_mean else 0 for i in range(len(numbers))]
print(numbers_gt_mean_ls)

[1, 1, 0, 0, 0, 1, 1, 1, 0]


7.Задано список `a` і елемент `element`. Перевірте програмно чи є в списку `a` елемент `element` з допомогою інструкцій `if ... else`. Виведіть на екран відповідь "Так, є", якщо цей елемент в списку присутній, або "Ні, нема", якщо його там немає.
Важливо це зробити без використання імпортів будь-яких бібліотек.

In [12]:
a = [64, 93, 60, 95, 66, 9, 15, 55, 77, 28, 74, 100, 13, 1, 85, 48, 66, 65, 59, 32]
element = 33

In [16]:
if element in a:
    print('Так,є')
else:
    print('Ні, нема')

Ні, нема


8.Тепер у нас не один елемент, а кілька. Вони наведені в змінній `elements`. За допомогою циклу і інструкцій `if ... else` для кожного елемента списку `elements` перевірте, чи він є в списку `a`. 
Виведіть на екран відповідь "Так, є елемент <вставити значення елемента>", якщо елемент в списку присутній, або "Ні, нема елемента <вставити значення елемента>", якщо його там немає.

In [17]:
elements = [44, 66, 77, 100]

In [18]:
for i in range(len(elements)):
    if elements[i] in a:
        print(f'Так, є елемент {elements[i]}')
    else:
        print(f'Ні, нема елемента {elements[i]}')

Ні, нема елемента 44
Так, є елемент 66
Так, є елемент 77
Так, є елемент 100


9.Заданий список `long`. З допомогою циклу створіть словник `n_repeats`, який містить інформацію "скільки разів повторюється кожен унікальний елемент списку `long`".
Наприклад:
```
long = [1, 1, 2, 2, 3]
# тоді `n_repeats` містить {1: 2, 2: 2, 3: 1}
```

Аби для кожного унікального елементу знайти, скільки разів він повторюється, виористайте метод списку `list.count(<елемент>)`. 
Виведіть `n_repeats` на екран. 

**Підказки**: 
1. Для початку може бути зручно отримати набір унікальних елементів списку.
2. Порядок ключів в словнику не має значення! Тобто ключі в словнику не обовʼязково мають бути в тому ж порядку, як наведено в прикладі результату. Але кількості зустрічань для кожного елемента мають співпадати з наведеними. 

In [19]:
long = [75, 39, 95, 1, 91, 43, 97, 24, 86, 75, 92, 91, 15, 15, 75, 95, 54, 29, 55, 98]

In [26]:
uniq_elements = list(set(long))
print(uniq_elements)

[1, 97, 98, 39, 75, 43, 15, 55, 86, 54, 24, 91, 92, 29, 95]


In [28]:
n_repeats = {}
for i in range(len(uniq_elements)):
    n_repeats[uniq_elements[i]] = long.count(uniq_elements[i])
print(n_repeats)
    
    

{1: 1, 97: 1, 98: 1, 39: 1, 75: 3, 43: 1, 15: 2, 55: 1, 86: 1, 54: 1, 24: 1, 91: 2, 92: 1, 29: 1, 95: 2}


Очікуваний результат: 
```
{1: 1, 97: 1, 98: 1, 39: 1, 75: 3, 43: 1, 15: 2, 55: 1, 86: 1, 54: 1, 24: 1, 91: 2, 92: 1, 29: 1, 95: 2}

```

10.Виконайте ту саму операцію, що і в попередньому заваднні, але використовуючи `dict comprehension`. Результат запишіть у змінну `n_repeats_dh` та результат виведіть на екран. 

In [45]:
n_repeats_dh = {uniq_elements[i]:long.count(uniq_elements[i]) for i in range(len(uniq_elements))}
print(n_repeats_dh)
                                                                                            

{1: 1, 97: 1, 98: 1, 39: 1, 75: 3, 43: 1, 15: 2, 55: 1, 86: 1, 54: 1, 24: 1, 91: 2, 92: 1, 29: 1, 95: 2}


11.Знайдіть три найменші **унікальні** елементи в списку `long` та виведіть їх на екран.

In [54]:
sort_list = sorted(uniq_elements)
print(sort_list[0:3])

[1, 15, 24]


Очікувана відповідь: [1, 15, 24]
